### ISSUE TITLE
Unable to reproduce Precision-Recall plot Supp. Fig. 5

### ISSUE THREAD

**Issue by rcedgar** [steineggerlab/foldseek-analysis#5]
![foldseek_prec_recall_screenshot](https://github.com/user-attachments/assets/ff422ca8-dc11-4276-990b-d79d86038955)

I have tried reproducing Supp. Fig. 5 using the scripts in this repository and also my own code but I get quite different results. For example, as shown in the figure above I find TM-align is better than DALI on Superfamily over the entire range, while your figure shows TM-align to be substantially worse (Fig. S5 on the left, my results on the right). My plot on the right was generated using your data and scripts as follows.

Hits downloaded from https://wwwuser.gwdguser.de/~compbiol/foldseek/scop.benchmark.result.tar.gz

<pre>
sort -rgk3 ../alns/TMalign.txt > TMalign.sorted.txt
sort -rgk3 ../alns/dali.txt > dali.sorted.txt
bench.fdr.noselfhit.awk TMalign.sorted.txt scop_lookup.fix.tsv <(cat $TMalign.sorted.txt) > TM-align.rocx
bench.fdr.noselfhit.awk dali.sorted.txt scop_lookup.fix.tsv <(cat dali.sorted.txt) > dali.rocx
</pre>

Plot column `PREC_SFAM` (y axis) vs. `RECALL_SFAM` (x axis).

As you can see, the plot for DALI looks right but TM-align looks very wrong. 

Any help in resolving this discrepancy will be much appreciated.

Also, the calculation of precision and recall in `bench.fdr.noselfhit.awk` appears to be making corrections compared to the standard formulas, but I don't understand how it works. Can you clarify? In particular, what is the variable `norm` is doing? Thanks!

**Comment by rcedgar** [comment]
After looking more closely at the source code, as best I can tell the calculations of both precision and recall are not correct.

Line 5 in bench.fdr.noselfhit.awk outputs 6 column headings:
<pre>
                    1           2           3            4             5             6
      print "PREC_FAM","PREC_SFAM","PREC_FOLD","RECALL_FAM","RECALL_SFAM","RECALL_FOLD";
</pre>

Line 37 in bench.fdr.noselfhit.awk outputs 7 columns:

<pre>
                                     1                     2                     3                 4                  5                  6       7
NR %1000 == 0{ print tp_fam/(tp_fam+fp), tp_sfam/(tp_sfam+fp), tp_fold/(tp_fold+fp), tp_fam / queries, tp_sfam / queries, tp_fold / queries, tp_fam; }
</pre>

Matching column headings to expressions in line 37:

<pre>
1. PREC_FAM     = tp_fam/(tp_fam+fp)
2. PREC_SFAM    = tp_sfam/(tp_sfam+fp)
3. PREC_FOLD    = tp_fold/(tp_fold+fp)

4. RECALL_FAM   = tp_fam / queries # queries = constant = 3566
5. RECALL_SFAM  = tp_sfam / queries
6. RECALL_FOLD  = tp_fold / queries
</pre>

Definitions of precision and recall are (see e.g. https://scikit-learn.org/1.5/auto_examples/model_selection/plot_precision_recall.html):

<pre>
# TP = number of true positive hits above the threshold
# FP = number of false positive hits above the threshold
# FN = number of false negatives at the threshold

Precision = TP/(TP + FP)
Recall = TP/(TP + FN)
</pre>

The calculation of precision appears to be wrong for family and superfamily because `fp` is always calculated for fold (see line 38, it is increased if `id2fold[$1] != id2fold[$2]`).

The calculation of recall appears to be wrong because the divisor should be (TP+FN) but is constant (always 3556).

**Comment by martin-steinegger** [comment]
Thanks for flagging this. We will have a look at it. But since the postdoc left the lab it might need some time. 
Regarding queries, we are computing a weighted ROC, meaning each query can contribute in total 1 if it recalls all possible TPs (TP=1/foldSize(norm)). Therefore, the number of queries is calculated as TP + FN. Precision is computed as expected: Precision = TP / (TP + FP) but we also apply weighting.

As for the TM-align results, we tried various methods to sort the hits. A reviewer recommended using the average of qTM and tTM scores, which indeed worked best. It’s possible that the uploaded file does not reflect this averaging.

**Comment by QY6616** [comment]
Do you emersion DALI and TM - align index, is to use the https://wwwuser.gwdguser.de/~compbiol/foldseek/scop.benchmark.result.tar.gz related script? Because I didn't find in the https://github.com/steineggerlab/foldseek-analysis DALI and TM - align related scripts  @rcedgar

**Comment by rcedgar** [comment]
@QY6616 sorry, I don't understand the question

**Comment by QY6616** [comment]
> [@QY6616](https://github.com/QY6616)抱歉，我不明白这个问题

@rcedgar  What I mean is: How did you reproduce the PR curve of DALI and TM-align? I also want to reproduce the PR curve of DALI and TM-align on the SCOPE dataset, but I haven't found the relevant scripts.

**Comment by rcedgar** [comment]
<pre>
foldseek-analysis/scopbenchmark/scripts/bench.fdr.noselfhit.awk
foldseek-analysis/scopbenchmark/scripts/bench.noselfhit.awk
foldseek-analysis/training/roc1.awk
</pre>

**Comment by rcedgar** [comment]
fyi the corresponding scripts & figures for the reseek paper are here  [https://github.com/rcedgar/reseek_bench](https://github.com/rcedgar/reseek_bench)

**Comment by QY6616** [comment]
> fyi the corresponding scripts & figures for the reseek paper are here https://github.com/rcedgar/reseek_bench

@rcedgar  Thank you very much for your help, but the link you provided cannot be opened. Could you please check it?

**Comment by rcedgar** [comment]
oops! that repo was still private, it should have been made public when the paper was published --  I fixed that, you should be able to reach it now.

**Comment by QY6616** [comment]
@rcedgar  It is accessible now! Actually, what I want to ask is: as shown in the figure, where can I download the ../data/scop.file.full.name.list and ./TM-align/TMalign files in foldseek-analysis? I see that you provided a download link: https://wwwuser.gwdguser.de/~compbiol/foldseek/scop.benchmark.result.tar.gz. Does this file contain the relevant files for TMalign?

<img width="835" alt="Image" src="https://github.com/user-attachments/assets/7ec4647e-581b-4fb3-a070-326824187961" />

**Comment by rcedgar** [comment]
Script `foldseek_rocx`.bash in `rcedgar/reseek_bench` runs the awk script from `steineggerlab/foldsee-analysis` , also `reseek_bench/awk` has a port of their awk code to python to make sure I understood the logic by ensuring that it generates the same results. I don't remember all the details, IIRC I had to make some changes and reconstruct some helper files, I don't think the code in `steineggerlab/foldsee-analysis` runs as-is. The `https://wwwuser.gwdguser.de/~compbiol/foldseek/scop.benchmark.result.tar.gz` download has alignments for DALI, TM-align and the other methods, I don't remember what else it has offhand.

**Comment by QY6616** [comment]
Thank you for your help, I will give it a try.

**Comment by QY6616** [comment]
@rcedgar Hello, I obtained the same PR curves for TMalign and DALI using the same method as you, but when I tried to reproduce the MMseq2 results, I encountered the data shown in the image. Did I overlook any details, or is there an issue with the data itself?

<img width="497" alt="Image" src="https://github.com/user-attachments/assets/04f48bd0-d718-4e04-a7d5-be66ea7d869b" />

**Comment by rcedgar** [comment]
The image seems to show a table of numbers, I don't really understand what you are asking. If you can't reproduce a figure from the Foldseek paper, then this seems to confirm the issue here. If so, I suggest it would be useful to show the MMseqs curve you obtained side-by-side to clarify the discrepancy, similar to the image I showed at the start of this issue. FWIW, I believe that some of the curves in the Foldseek paper are wrong because some curves are reproduced by their code but others are not; this could possibly be explained by something like a copy/paste error. The TM-align example is striking because the published figure implies that TM-align is much worse than DALI, while I find that TM-align is better than DALI using both the code here and also my own code.
